languages/ruby/spec/oso/polar/data_filtering/new_filter_spec.rb

```diff
@@ -0,0 +1,419 @@
+            allow(sign: Sign{planet}, "read", _: Person{sign}) if planet.name = "jupiter";
+            # every sign can read a pisces named sam
+            allow(_: Sign, "read", _: Person {sign, name: "sam"}) if sign.name = "pisces";
+            # earth signs can read people with air signs
+            allow(_: Sign{element: "earth"}, "read", _: Person{sign}) if sign.element = "air";
+          POL
+
+          test = lambda do |person, sign|
+            (person.sign == sign && sign.planet.name == 'jupiter') ||
+              (person.name == 'sam' && person.sign.name == 'pisces') ||
+              (sign.element == 'earth' && person.sign.element == 'air')
+          end
+          Sign.all.each do |sign|
+            expected = Person.all.select do |person|
+              test[person, sign]
+            end
+            query = subject.authzd_query sign, 'read', Person
+            expect(query.to_a).to contain_exactly(*expected)
+          end
+        end
+
+        it 'test_nested_relationship_many_single' do
+          subject.load_str <<~POL
+            allow(_: Person{sign}, _, _: Planet{signs}) if sign in signs;
+          POL
+          Person.all.each do |person|
+            query = subject.authzd_query person, nil, Planet
+            expect(query.to_a).to eq [person.sign.planet]
+          end
+        end
+
+        it 'test_nested_relationship_many_many' do
+          subject.load_str <<~POL
+            allow(person: Person, _, _: Planet{signs}) if
+              person in signs.people;
+          POL
+
+          Person.all.each do |person|
+            query = subject.authzd_query person, nil, Planet
+            expect(query.to_a).to eq [person.sign.planet]
+          end
+        end
+
+        it 'test_nested_relationship_many_many_constrained' do
+          subject.load_str <<~POL
+            allow(person: Person{name: "eden"}, _, _: Planet{signs}) if
+              sign in signs and
+              person in sign.people;
+          POL
+
+          Person.all.each do |person|
+            query = subject.authzd_query person, nil, Planet
+            if person == eden
+              expect(query.to_a).to eq [person.sign.planet]
+            else
+              expect(query.to_a).to eq []
+            end
+          end
+        end
+
+        it 'test_partial_in_collection' do
+          subject.load_str(
+            'allow(_: Planet{signs}, _, sign: Sign) if sign in signs;'
+          )
+          Planet.all.each do |planet|
+            query = subject.authzd_query planet, nil, Sign
+            expect(query.to_a).to contain_exactly(*planet.signs)
+          end
+        end
+
+        it 'test_empty_constraints_in' do
+          subject.load_str <<~POL
+            allow(_, _, _: Planet{signs}) if
+              _ in signs;
+          POL
+          query = subject.authzd_query 'gwen', 'get', Planet
+          expect(query.to_a).to contain_exactly(*Planet.where.not(name: 'pluto'))
```

Comment:
Might be good to do some verification of SQL generated in some of these like we do in the SQLAlchemy library to make sure that generation doesn't add any spurious stuff.

Reply:
(Not now, just a suggestion if you're writing more tests)